In [1]:
import pandas as pd
import numpy as np
# sklearn.metrics import mean_absolute_error,mean_squared_error

In [2]:
user_ratings = pd.read_excel('./Metric-SpreadSheet-Assignment.xlsx',skip_footer=70,skiprows=1,index_col=0)

In [3]:
user_ratings = user_ratings.iloc[:,:-3]

In [4]:
user_predictings = pd.read_excel('./Metric-SpreadSheet-Assignment.xlsx',skip_footer=56,skiprows=15,index_col=0)

In [5]:
user_predictings = user_predictings.iloc[:,:-3]

In [6]:
user_predictings.columns = user_ratings.columns

In [7]:
MAE_by_user = pd.read_excel('./Metric-SpreadSheet-Assignment.xlsx',skiprows=27,skip_footer=44,index_col=0)

In [8]:
MAE_by_user = MAE_by_user.iloc[:,:-3]

In [9]:
MAE_by_user.columns = user_ratings.columns

# 1.Compute the mean absolute error for each user

In [10]:
MAE_by_user['MAE'] = MAE_by_user.mean(axis=1)

In [11]:
MAE_by_user.sort_values('MAE',ascending=False)

11: Star Wars: Episode IV - A New Hope (1977)  12: Finding Nemo (2003)  \
Users                                                                           
2267                                             NaN                      NaN   
3556                                             0.3                      0.7   
5261                                             NaN                      NaN   
918                                              1.0                      2.7   
860                                              0.0                      1.0   
4942                                             NaN                      1.1   
5136                                             0.4                      0.7   
442                                              0.6                      NaN   
2492                                             0.4                      0.8   
2824                                             0.2                      NaN   

       13: Forrest Gump (1994)  14: American Beauty (1999)  \
Users                                                        
2267                       0.4                         NaN   
3556                       0.3                         2.0   
5261                       0.4                         NaN   
918                        0.1                         NaN   
860                        0.7                         NaN   
4942                       0.7                         1.0   
5136                       0.3                         0.3   
442                        1.2                         1.0   
2492                       0.7                         0.8   
2824                       0.2                         NaN   

       22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)  \
Users                                                                      
2267                                                 NaN                   
3556                                                 1.4                   
5261                                                 2.0                   
918                                                  0.4                   
860                                                  1.6                   
4942                                                 0.7                   
5136                                                 1.2                   
442                                                  1.2                   
2492                                                 0.7                   
2824                                                 0.6                   

       24: Kill Bill: Vol. 1 (2003)  \
Users                                 
2267                            0.6   
3556                            0.9   
5261                            1.2   
918                             NaN   
860                             0.6   
4942                            0.6   
5136                            0.2   
442                             0.8   
2492                            0.7   
2824                            0.2   

       38: Eternal Sunshine of the Spotless Mind (2004)  \
Users                                                     
2267                                                1.1   
3556                                                NaN   
5261                                                NaN   
918                                                 0.6   
860                                                 NaN   
4942                                                0.6   
5136                                                1.8   
442                                                 0.8   
2492                                                NaN   
2824                                                NaN   

       63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)  77: Memento (2000)  \
Users                                                                      
2267                                          

# 2. Also compute the mean absolute error associated with the predictions for each movie. For example, Forrest Gump has an MAE of 0.5.

In [12]:
MAE_by_user.mean(axis=0).sort_values(ascending=False)

12: Finding Nemo (2003)                                                               1.166667
22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)                     1.088889
14: American Beauty (1999)                                                            1.020000
38: Eternal Sunshine of the Spotless Mind (2004)                                      0.980000
MAE                                                                                   0.731107
77: Memento (2000)                                                                    0.662500
24: Kill Bill: Vol. 1 (2003)                                                          0.644444
13: Forrest Gump (1994)                                                               0.500000
63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)                                         0.460000
11: Star Wars: Episode IV - A New Hope (1977)                                         0.414286
85: Raiders of the Lost Ark (Indiana Jones and the

# 3. Compute the overall MAE three ways (three decimal places is enough):

In [13]:
movie_names = ['11: Star Wars: Episode IV - A New Hope (1977)',
       '12: Finding Nemo (2003)', '13: Forrest Gump (1994)',
       '14: American Beauty (1999)',
       '22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)',
       '24: Kill Bill: Vol. 1 (2003)',
       '38: Eternal Sunshine of the Spotless Mind (2004)',
       '63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)', '77: Memento (2000)',
       '85: Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)',
]

In [14]:
print('normal MAE',MAE_by_user[movie_names].sum().sum()/MAE_by_user[movie_names].count().sum())

normal MAE 0.7246376811594202


In [15]:
print('by user MAE',MAE_by_user.MAE.mean())

by user MAE 0.731107142857143


In [16]:
print('by movie MAE',MAE_by_user[movie_names].mean(axis=0).mean())

by movie MAE 0.7316785714285714


# 4. Next, create a matrix with the squared errors (instead of absolute errors), and compute the MSE and RMSE (root-mean-squared-error) for each user. 

In [17]:
SE_by_user = (user_predictings[movie_names] - user_ratings[movie_names]).applymap(lambda x:np.square(x))

In [18]:
SE_by_user['MSE'] = SE_by_user.mean(axis=1)
SE_by_user['RMSE'] = SE_by_user['MSE'].apply(lambda x:np.sqrt(x))

In [19]:
SE_by_user.sort_values('MSE',ascending=False)

11: Star Wars: Episode IV - A New Hope (1977)  12: Finding Nemo (2003)  \
Users                                                                           
918                                             1.00                     7.29   
5261                                             NaN                      NaN   
3556                                            0.09                     0.49   
2267                                             NaN                      NaN   
860                                             0.00                     1.00   
5136                                            0.16                     0.49   
442                                             0.36                      NaN   
4942                                             NaN                     1.21   
2492                                            0.16                     0.64   
2824                                            0.04                      NaN   

       13: Forrest Gump (1994)  14: American Beauty (1999)  \
Users                                                        
918                       0.01                         NaN   
5261                      0.16                         NaN   
3556                      0.09                        4.00   
2267                      0.16                         NaN   
860                       0.49                         NaN   
5136                      0.09                        0.09   
442                       1.44                        1.00   
4942                      0.49                        1.00   
2492                      0.49                        0.64   
2824                      0.04                         NaN   

       22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)  \
Users                                                                      
918                                                 0.16                   
5261                                                4.00                   
3556                                                1.96                   
2267                                                 NaN                   
860                                                 2.56                   
5136                                                1.44                   
442                                                 1.44                   
4942                                                0.49                   
2492                                                0.49                   
2824                                                0.36                   

       24: Kill Bill: Vol. 1 (2003)  \
Users                                 
918                             NaN   
5261                           1.44   
3556                           0.81   
2267                           0.36   
860                            0.36   
5136                           0.04   
442                            0.64   
4942                           0.36   
2492                           0.49   
2824                           0.04   

       38: Eternal Sunshine of the Spotless Mind (2004)  \
Users                                                     
918                                                0.36   
5261                                                NaN   
3556                                                NaN   
2267                                               1.21   
860                                                 NaN   
5136                                               3.24   
442                                                0.64   
4942                                               0.36   
2492                                                NaN   
2824                                                NaN   

       63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)  77: Memento (2000)  \
Users                                                                      
918                                           

In [20]:
SE_by_user.sort_values('RMSE',ascending=False)

11: Star Wars: Episode IV - A New Hope (1977)  12: Finding Nemo (2003)  \
Users                                                                           
918                                             1.00                     7.29   
5261                                             NaN                      NaN   
3556                                            0.09                     0.49   
2267                                             NaN                      NaN   
860                                             0.00                     1.00   
5136                                            0.16                     0.49   
442                                             0.36                      NaN   
4942                                             NaN                     1.21   
2492                                            0.16                     0.64   
2824                                            0.04                      NaN   

       13: Forrest Gump (1994)  14: American Beauty (1999)  \
Users                                                        
918                       0.01                         NaN   
5261                      0.16                         NaN   
3556                      0.09                        4.00   
2267                      0.16                         NaN   
860                       0.49                         NaN   
5136                      0.09                        0.09   
442                       1.44                        1.00   
4942                      0.49                        1.00   
2492                      0.49                        0.64   
2824                      0.04                         NaN   

       22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)  \
Users                                                                      
918                                                 0.16                   
5261                                                4.00                   
3556                                                1.96                   
2267                                                 NaN                   
860                                                 2.56                   
5136                                                1.44                   
442                                                 1.44                   
4942                                                0.49                   
2492                                                0.49                   
2824                                                0.36                   

       24: Kill Bill: Vol. 1 (2003)  \
Users                                 
918                             NaN   
5261                           1.44   
3556                           0.81   
2267                           0.36   
860                            0.36   
5136                           0.04   
442                            0.64   
4942                           0.36   
2492                           0.49   
2824                           0.04   

       38: Eternal Sunshine of the Spotless Mind (2004)  \
Users                                                     
918                                                0.36   
5261                                                NaN   
3556                                                NaN   
2267                                               1.21   
860                                                 NaN   
5136                                               3.24   
442                                                0.64   
4942                                               0.36   
2492                                                NaN   
2824                                                NaN   

       63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)  77: Memento (2000)  \
Users                                                                      
918                                           

# 5. Then, compute the overall MSE and RMSE across all predictions (just the normal way, no need to compute by-user or by-movie). Again, remember to average all the squared errors to compute MSE, then take the square root of the MSE to get RMSE.

In [21]:
print('overall MSE',SE_by_user[movie_names].sum().sum()/SE_by_user[movie_names].count().sum())

overall MSE 0.7944927536231883


In [22]:
print('overall RMSE',np.sqrt(SE_by_user[movie_names].sum().sum()/SE_by_user[movie_names].count().sum()))

overall RMSE 0.891343229976


# 6. Next, compute the correlation between ratings and predictions using the built-in CORREL() function. First, do this for each user (e.g., the correlation between ratings and predictions for user 5136 is 0.1698 -- only about 17%, and therefore not very good). Then, apply the same correlation to the full matrices.

In [23]:
user_predictings[movie_names].corrwith(user_ratings[movie_names],axis=1).sort_values(ascending=False)

Users
5261    0.920746
442     0.774450
2267    0.633743
918     0.458946
2492    0.413615
4942    0.347857
2824    0.347677
5136    0.169813
3556   -0.095887
860    -0.148181
dtype: float64

In [24]:
user_predictings[movie_names].unstack().corr(user_ratings[movie_names].unstack(),)

0.56644751480428168